src/telemetry/metric_registry.cc

```diff
@@ -185,6 +185,82 @@ class impl_base : public metric_registry_impl {
+  // rather than using some other handle.
+  class ProxyCollector {
+  public:
+    ProxyCollector(broker::telemetry::metrics_collector& collector)
+      : collector_(collector){};
+    void operator()(const ct::metric_family* family, const ct::metric* instance,
+                    const ct::dbl_counter* counter) {
+      auto bt_family = reinterpret_cast<const dbl_counter_family_hdl*>(family);
+      auto bt_hdl = reinterpret_cast<const dbl_counter_hdl*>(counter);
+      collector_(bt_family, bt_hdl, _labels(instance));
+    }
+
+    void operator()(const ct::metric_family* family, const ct::metric* instance,
+                    const ct::int_counter* counter) {
+      auto bt_family = reinterpret_cast<const int_counter_family_hdl*>(family);
+      auto bt_hdl = reinterpret_cast<const int_counter_hdl*>(counter);
+      collector_(bt_family, bt_hdl, _labels(instance));
+    }
+
+    void operator()(const ct::metric_family* family, const ct::metric* instance,
+                    const ct::dbl_gauge* gauge) {
+      auto bt_family = reinterpret_cast<const dbl_gauge_family_hdl*>(family);
+      auto bt_hdl = reinterpret_cast<const dbl_gauge_hdl*>(gauge);
+      collector_(bt_family, bt_hdl, _labels(instance));
+    }
+
+    void operator()(const ct::metric_family* family, const ct::metric* instance,
+                    const ct::int_gauge* gauge) {
+      auto bt_family = reinterpret_cast<const int_gauge_family_hdl*>(family);
+      auto bt_hdl = reinterpret_cast<const int_gauge_hdl*>(gauge);
+      collector_(bt_family, bt_hdl, _labels(instance));
+    }
+
+    void operator()(const ct::metric_family* family, const ct::metric* instance,
+                    const ct::dbl_histogram* val) {
+      auto bt_family
+        = reinterpret_cast<const dbl_histogram_family_hdl*>(family);
+      auto bt_hdl = reinterpret_cast<const dbl_histogram_hdl*>(val);
+      collector_(bt_family, bt_hdl, _labels(instance));
+    }
+
+    void operator()(const ct::metric_family* family, const ct::metric* instance,
+                    const ct::int_histogram* val) {
+      auto bt_family
+        = reinterpret_cast<const int_histogram_family_hdl*>(family);
+      auto bt_hdl = reinterpret_cast<const int_histogram_hdl*>(val);
+      collector_(bt_family, bt_hdl, _labels(instance));
+    }
+
+  private:
+    std::vector<std::string_view> _labels(const ct::metric* instance) {
+      std::vector<std::string_view> r;
+      for (const auto& l : instance->labels()) {
+        const auto& v = l.value();
+        // Not super sure this is great: The std::string_view points at
+        // the caf memory of the caf::string_view. Fingers-crossed.
+        r.push_back(std::string_view(v.data(), v.size()));
+      }
+      return r;
+    }
+
+    metrics_collector& collector_;
+  };
+
+  void collect(metrics_collector& collector) {
+    auto proxy_collector = ProxyCollector(collector);
+    reg_->collect(proxy_collector);
+  }
```

Comment:
The `operator()` overloads invoke UB, e.g., by using a `reinterpret_cast` from `ct::metric_family` to `dbl_counter_family_hdl` (which is the type-erased version of `ct::metric_family_impl<dbl_counter>`). This "works" on most current platforms, but only if there arent' any virtual base types involved.

In any case, the proxy class is indeed unnecessary and adds a lot of boilerplate code. Could just use a couple of free function helpers and a lambda:

```cpp
std::vector<std::string_view> extract_label_values(const ct::metric* instance) {
  auto get_value = [](const auto& lbl) {
    auto val = lbl.value();
    return std::string_view{val.data(), val.size()};
  };
  auto labels = instance->labels();
  std::vector<std::string_view> result;
  std::transform(labels.begin(), labels.end(), std::back_inserter(result), get_value);
  return result;
}
    
const auto* opaque(const ct::metric_family_impl<ct::dbl_counter>* family) {
  return reinterpret_cast<const dbl_counter_hdl*>(family);
}

// .. some more opaque overloads, could also be macro-generated ..
```

With these, the `collect` function becomes straightforward:

```cpp
  void collect(metrics_collector& collector) {
    auto fn = [&collector](const auto* family, const ct::metric* instance, const auto* obj) {
      auto labels = extract_label_values(instance);
      collector(opaque(family), opaque(obj), labels);
    };
    reg_->collect(fn);
  }
```

(Please also note my comment regarding extracting the label values only. So probably the free function would be `extract_labels` and assemble the key-value pairs. It's probably also a good idea to pass the vector the `extract` function and re-fill the same vector over and over again instead of allocating fresh memory each time.)

Reply:
I've followed your suggested route with using lambda and free-functions, but just for clarification:

> The operator() overloads invoke UB, e.g., by using a reinterpret_cast from ct::metric_family to dbl_counter_family_hdl (which is the type-erased version of ct::metric_family_impl<dbl_counter>). This "works" on most current platforms, but only if there arent' any virtual base types involved.

This would have been "better" if the individual `operator()` implementations had used concrete `ct::dbl_counter_family *` rather than `ct::metric_family *` (as the `opaque()` helpers now will take?).

The docs and the prom collector which I had copied use the `ct::metric_family *` - that's where this was coming from:
https://github.com/actor-framework/actor-framework/blob/master/manual/Metrics.rst#the-metric-registry

Reply:
I've pushed the changes. Thanks, it got shorter and looks better.

TBH, I'm slightly intimidated by that lambda auto type deduction ;-)

For the final merge I'll squash the commits together.

Mind giving it another look? Thanks!

Reply:
> This would have been "better" if the individual `operator()` implementations had used concrete `ct::dbl_counter_family *` rather than `ct::metric_family *` (as the `opaque()` helpers now will take?).

Yes. By taking the argument as `ct::metric_family*`, you're forcing a downcast to the base type. For going back to the concrete type, you need to use `static_cast` (if you're 100% certain you're casting to the right type) or `dynamic_cast`. Using a `reinterpret_cast` to do this is unsafe.

> The docs and the prom collector which I had copied use the ct::metric_family * - that's where this was coming from:

The CAF example does not care about the concrete type of the family objects. You generally only need those for having access to `get_or_add` (i.e., navigate to or create new instances), but while you are iterating the registry, you'll have `const` access anyway. When collecting metrics, the base type is usually all you need for the family. The new `metrics_collector` could've done the same, but then the `operator()` overloads should take `metric_family_hdl*` as first argument in all overloads. The question really is whether you see a use case for having the concrete family available or not. If not, you could drop half the `opaque` overloads (for the family types) and streamline the code a bit. :)

Reply:
I've did as you suggested. The Zeek consumer only did upcast. I did keep the const `upcast()` changes in this PR. If you'd like to see them go, can remove them.

I've addressed the other two nitpicks as well.

Thanks again.

Reply:
Ah - I'm not sure what more streamlining there could've been done.

Reply:
Sorry, that was unclear. I just meant getting rid of the macro stuff for the family overloads to streamline the code. Looks good as-is now.